misc/inet/inet.1
.\" Copyright (c) 1986 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	@(#)nslookup.1	1.2 (Berkeley) 11/21/87
.\"
.TH INET 1 "August 8, 1988"
.SH NAME
inet \- ask the domain name system for a host name or address
.SH SYNOPSIS
.B inet
[ -d ]
[ -v ]
[ -h ]
[ -s nameserver[:port] ]
.I hostname\ |\ host-IP-address
\|...
.SH DESCRIPTION
.IR Inet
queries DARPA Internet domain name servers. 
Given a (list of) host names, it uses
.IR gethostbyname (3)
to find the canonical host name and all listed interfaces;
given host addresses, it calls
.IR gethostbyaddr (3)
to find the canonical host name.
It uses the domain resolver library to do this, bypassing any /etc/hosts
or Sun Yellow Pages tables.
.sp
Since it is linked with the BIND 4.8 domain resolver library, it
applies the sliding-match host name lookup described in resolver(3).
As described there, a trailing ``.'' on a host name suppresses the
sliding match search, forcing a lookup on just the specified name.
.sp
The -d option enables debugging (setting the RES_DEBUG flag mentioned in
.IR resolver (3)
); details of all domain requests sent and responses received are printed.
.sp
The -v option selects using a virtual circuit (TCP stream) rather than
UDP datagrams for sending queries.  It is worth using when asking about
many hosts at a time.
.sp
The -h option forces a name to be looked up as a host name even if
it appears to be a numeric IP address.
.sp
The -n option prints out only the DNS name of the given address(es) or name(s).
.sp
The -i option prints out only the IP address(es) of the given name(s) or address(es).
.sp
The -s option selects the name server(s) to query rather than the ones
listed in /etc/resolv.conf.  Multiple -s options are allowed; they are
queried in the order given until a response is received.
Suffixing ``:'' and a decimal port number to a name server name/IP-address
directs queries to that port rather than the standard 53.
.sp
.SH DIAGNOSTICS
.in +4
.ti -4
HOST_NOT_FOUND -- inet received an authoritative indication that the
host or IP address is not known anywhere in the domain system
.sp
.ti -4
TRY_AGAIN -- either no response was recieved or the response was not
authoritative.  Another try might yield a definite answer.
This error can also appear if the server(s) are not working correctly.
.sp
.ti -4
NO_DATA -- the host exists but no IP address is known; there may be
other information such as MX mail-forwarding records.
.sp
.ti -4
NO_RECOVERY -- a server complained that inet had sent an invalid request;
should not happen.
.in -4
.sp
.SH FILES
/etc/resolv.conf	default domain name and name server addresses
.SH SEE ALSO
nslookup(1), nstest(1), resolver(3), resolver(5)


misc/inet/inet.c
/* Inet program, who knows where we got it... */

#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netdb.h>
#if	BSD >= 43
#include <arpa/inet.h>
#else	/* BSD <= 42 */
 /* 4.2 and SUN arpa/inet.h misdeclare inet_addr. */
extern unsigned long inet_addr();
#endif	/* BSD <= 42 */
#include <ctype.h>
#include <arpa/nameser.h>
#include <resolv.h>
#ifndef NAMESERVER_PORT
#define NAMESERVER_PORT 53
#endif

/* ------------------------------------------------------------------------- */
extern char    *inet_ntoa();
// extern char    *index(char *, char);
#include <strings.h>
// extern int	fprintf();
// extern int	printf();
// extern void	bcopy(char *, char *, int);
// extern void	bzero(char *, int);

extern int      errno;
extern int      h_errno;
/* ------------------------------------------------------------------------- */
char           *progname = "(Unknown)";


char           *h_errors[] = {
  "no error", "HOST_NOT_FOUND", "TRY_AGAIN", "NO_RECOVERY", "NO_DATA"
};

/* ------------------------------------------------------------------------- */
int             main(argc, argv)
int             argc;
char           *argv[];
{
  int    i;
  unsigned long   in;
  struct sockaddr_in inet_sin;
  struct hostent *hp;
  char  *p;
  char           *pport;
  int             ishost = 0;
  int             altns = 0;
  int		  nameonly = 0;
  int		  iponly = 0;
  int		  flag;

  if (argc > 0)
    progname = argv[0];
  if (argc < 2)
    goto Usage;
  if (res_init() < 0) {
    (void) fprintf(stderr, "%s: res_init() failed.\n", progname);
    exit(1);
  }
  hp = NULL;
  for (i = 1; i < argc; i++) {
    if (argv[i][0] == '-') {
      switch (argv[i][1]) {
	case 'v':			/* use virtual circuit */
	  _res.options |= RES_USEVC;
	  break;
	case 'd':			/* enable resolver debug output */
	  _res.options |= RES_DEBUG;
	  break;
	case 's':
	  if (*(p = &(argv[i][2])) == '\0')
	    p = argv[++i];
	  if ((pport = index(p, ':')) != NULL)
	    *pport++ = '\0';
	  bzero((char *)(&inet_sin), sizeof(inet_sin));
	  inet_sin.sin_family = AF_INET;
	  if ((inet_sin.sin_addr.s_addr = inet_addr(p)) == (unsigned int)-1) {
	    hp = gethostbyname(p);
	    if (hp == NULL) {
	      (void) fprintf(stderr,
			     "%s: unknown host\n", p);
	      exit(2);
	    }
	    bcopy(hp->h_addr, (char *)(&inet_sin.sin_addr),
		  hp->h_length);
	  }
	  if (pport != NULL)
	    inet_sin.sin_port = htons(atoi(pport));
	  if (inet_sin.sin_port == 0)
	    inet_sin.sin_port = htons(NAMESERVER_PORT);
	  if (altns == 0) {
	    altns++;
	    _res.nscount = 0;
	  }
	  if (_res.nscount >= MAXNS) {
	    (void) fprintf(stderr,
			   "Can only specify %d name servers\n",
			   MAXNS);
	    exit(3);
	  }
	  _res.nsaddr_list[_res.nscount++] = inet_sin;
	  break;
	case 'h':
	  ishost++;
	  break;
	case 'n':
	  nameonly = 1;
	  iponly = 0;
	  break;
	case 'i':
	  iponly = 1;
	  nameonly = 0;
	  break;
	default:
	Usage:
	  (void) fprintf(stderr, "\
Usage: %s [-d] [-v] [-h] [-s nameserver[:port]] host-name-or-address ...\n\
	-d Enable debug (display queries and responses)\n\
	-v Use virtual circuit (TCP rather than UDP)\n\
	-h Force lookup as host name even if it looks like IP address\n\
	-n Print out names, not IP address(es)\n\
	-i Print out IP address(es), not names\n\
	-s Ask that nameserver rather than those in /etc/resolv.conf.\n\
",
			 progname);
	  exit(1);
      }
      continue;
    }
    if (ishost || isalpha(argv[i][0])) {
      hp = gethostbyname(argv[i]);
    } else if ((in = inet_addr(argv[i])) != (unsigned int)-1) {
      hp = gethostbyaddr((char *)(&in), sizeof(in), AF_INET);
    }
    if (nameonly == 0 && iponly == 0) {
      printf("%s: ", argv[i]);
    }
    if (hp != NULL) {
      if (nameonly == 1) {		/* if only name, print it */
	printf("%s", hp->h_name);
      } else {
	flag = 0;			/* if need to print a leading space */
	if (iponly == 0) {		/* if both */
	  printf("host %s <=> ", hp->h_name);
	}
	while (*hp->h_addr_list != NULL) {
	  bcopy(hp->h_addr_list[0], (char *) &in, hp->h_length);
	  inet_sin.sin_addr.s_addr = in;
	  if (flag != 0) {
	    printf(" ");
	  }
	  printf("%s", inet_ntoa(inet_sin.sin_addr));
	  hp->h_addr_list++;
	  flag = 1;
	}
      }
      printf("\n");
    } else {
      printf("unknown (%s)\n", h_errors[h_errno]);
    }
  }
  return(0);
}

/* ------------------------------------------------------------------------- */
/* End of file inet.c */
